File: hena/tyama_hena24_enum.m
//usr/bin/true; tmpfile=$(mktemp); gcc -O2 -o $tmpfile -framework Foundation "$0" && $tmpfile; rm -f $tmpfile; exit

// http://qiita.com/Nabetani/items/1c83005a854d2c6cbb69
// http://nabetani.sakura.ne.jp/hena/ord24eliseq/

#import <stdio.h>
#import <Foundation/Foundation.h>

int isqrt(int n){
	if(n<=0)return 0;
	if(n<4)return 1;
	int x=0,y=n;
	for(;x!=y&&x+1!=y;)x=y,y=(n/y+y)/2;
	return x;
}
int icbrt(int n){
	if(n<0)return -icbrt(-n);
	if(n==0)return 0;
	if(n<8)return 1;
	int x=0,y=n;
	for(;x!=y&&x+1!=y;)x=y,y=(n/y/y+y*2)/3;
	return x;
}

typedef bool(*tbl)(long long);
typedef bool(*tbll)(long long,long long);

@interface generate_impl:NSEnumerator
-init;
-(NSNumber*)nextObject;
@end
@interface drop_prev_impl:NSEnumerator
-init:(tbl)check _:(id)prev;
-(NSNumber*)nextObject;
@end
@interface drop_next_impl:NSEnumerator
-init:(tbl)check _:(id)prev;
-(NSNumber*)nextObject;
@end
@interface drop_n_impl:NSEnumerator
-init:(tbll)check _:(long long)n _:(id)prev;
-(NSNumber*)nextObject;
@end

@implementation generate_impl
{
	long long i;
}
-init
{
	i=0;
	[super init];
	return self;
}
-(NSNumber*)nextObject;
{
	i+=1;
	return [NSNumber numberWithLongLong:i];
}
@end
id generate(){return [[[generate_impl alloc] init] autorelease];}

@implementation drop_prev_impl
{
	id prev;
	tbl check;
	long long a;
	long long b;
}
-init:(tbl)check_ _:(id)prev_
{
	prev=prev_;
	check=check_;
	a=0;
	b=[[prev nextObject] longLongValue];
	[super init];
	return self;
}
-(NSNumber*)nextObject;
{
	for(;;){
		a=b;
		b=[[prev nextObject] longLongValue];
		if(!check(b)){
			return [NSNumber numberWithLongLong:a];
		}
	}
}
@end
id drop_prev(tbl check,id prev){return [[[drop_prev_impl alloc] init:check _:prev] autorelease];}

@implementation drop_next_impl
{
	id prev;
	tbl check;
	bool first;
	long long a;
	long long b;
}
-init:(tbl)check_ _:(id)prev_
{
	prev=prev_;
	check=check_;
	first=true;
	a=0;
	b=0;
	[super init];
	return self;
}
-(NSNumber*)nextObject;
{
	for(;;){
		a=b;
		b=[[prev nextObject] longLongValue];
		if(first||!check(a)){
			first=false;
			return [NSNumber numberWithLongLong:b];
		}
	}
}
@end
id drop_next(tbl check,id prev){return [[[drop_next_impl alloc] init:check _:prev] autorelease];}

@implementation drop_n_impl
{
	id prev;
	tbll check;
	long long n;
	long long i;
}
-init:(tbll)check_ _:(long long)n_ _:(id)prev_
{
	prev=prev_;
	check=check_;
	n=n_;
	i=0;
	[super init];
	return self;
}
-(NSNumber*)nextObject;
{
	for(;;){
		i+=1;
		long long a=[[prev nextObject] longLongValue];
		if(!check(i,n)){
			return [NSNumber numberWithLongLong:a];
		}
	}
}
@end
id drop_n(tbll check,long long n,id prev){return [[[drop_n_impl alloc] init:check _:n _:prev] autorelease];}

bool is_sq(long long n){
	long long x=isqrt(n);
	return x*x==n;
}
bool is_cb(long long n){
	long long x=icbrt(n);
	return x*x*x==n;
}
bool is_multiple(long long i,long long n){return i%n==0;}
bool is_le(long long i,long long n){return i<=n;}

int main(){
	for(;;){
		id stream=generate();
		for(;;){
			char c=fgetc(stdin);
			if(c=='\n')break;
			else if(c=='S')stream=drop_next(is_sq,stream);
			else if(c=='s')stream=drop_prev(is_sq,stream);
			else if(c=='C')stream=drop_next(is_cb,stream);
			else if(c=='c')stream=drop_prev(is_cb,stream);
			else if(c=='h')stream=drop_n(is_le,100,stream);
			else if(c=='2')stream=drop_n(is_multiple,2,stream);
			else if(c=='3')stream=drop_n(is_multiple,3,stream);
			else if(c=='4')stream=drop_n(is_multiple,4,stream);
			else if(c=='5')stream=drop_n(is_multiple,5,stream);
			else if(c=='6')stream=drop_n(is_multiple,6,stream);
			else if(c=='7')stream=drop_n(is_multiple,7,stream);
			else if(c=='8')stream=drop_n(is_multiple,8,stream);
			else if(c=='9')stream=drop_n(is_multiple,9,stream);
			else return 0;
		}
		int ctr=0;
		for(NSNumber *num in stream){
			if(ctr>0)putchar(',');
			printf("%lld",[num longLongValue]);
			if(++ctr==10)break;
		}
		puts("");
		fflush(stdout);
	}
}

File: hena/tyama_hena24_enum.clj
#!/usr/bin/env clojure
;http://nabetani.sakura.ne.jp/hena/ord24eliseq/

(defn expt_impl [r n k]
	(if (<= k 0) r (recur (* r (if (> (mod k 2) 0) n 1)) (* n n) (quot k 2)))
)

(defn expt [n k]
	(expt_impl 1 n k)
)

(defn isqrt_impl [n x y]
	(if (or (= x y) (= (+ x 1) y)) x
		(recur n y (quot (+ (quot n y) y) 2))
	)
)

(defn isqrt [n]
	(if (<= n 0) 0
		(if (< n 4) 1
			(isqrt_impl n 0 n)
		)
	)
)

(defn icbrt_impl [n x y]
	(if (or (= x y) (= (+ x 1) y)) x
		(recur n y (quot (+ (quot (quot n y) y) (* y 2)) 3))
	)
)

(defn icbrt [n]
	(if (< n 0) (- 0 (icbrt (- 0 n)))
		(if (= n 0) 0
			(if (< n 8) 1
				(icbrt_impl n 0 n)
			)
		)
	)
)

(defn generate []
	(let [i (atom 0)]
		(fn []
			(reset! i (+ @i 1))
			@i
		)
	)
)

(defn drop_prev [check prev]
	(let [a (atom 0) b (atom (prev))]
		(fn []
			(reset! a @b)
			(reset! b (prev))
			(while (check @b)
				(reset! a @b)
				(reset! b (prev))
			)
			@a
		)
	)
)

(defn drop_next [check prev]
	(let [a (atom 0) b (atom 0) first (atom true)]
		(fn []
			(reset! a @b)
			(reset! b (prev))
			(while (and (not @first) (check @a))
				(reset! a @b)
				(reset! b (prev))
			)
			(reset! first false)
			@b
		)
	)
)

(defn drop_n [check n prev]
	(let [i (atom 0) a (atom 0)]
		(fn []
			(reset! i (+ @i 1))
			(reset! a (prev))
			(while (check @i n)
				(reset! i (+ @i 1))
				(reset! a (prev))
			)
			@a
		)
	)
)

(defn is_sq [n]
	(= (expt (isqrt n) 2) n)
)
(defn is_cb [n]
	(= (expt (icbrt n) 3) n)
)
(defn is_multiple [i n]
	(= (mod i n) 0)
)
(defn is_le [i n]
	(<= i n)
)

(def f {
	\S (fn [prev] (drop_next is_sq prev))
	\s (fn [prev] (drop_prev is_sq prev))
	\C (fn [prev] (drop_next is_cb prev))
	\c (fn [prev] (drop_prev is_cb prev))
	\h (fn [prev] (drop_n is_le 100 prev))
	\2 (fn [prev] (drop_n is_multiple 2 prev))
	\3 (fn [prev] (drop_n is_multiple 3 prev))
	\4 (fn [prev] (drop_n is_multiple 4 prev))
	\5 (fn [prev] (drop_n is_multiple 5 prev))
	\6 (fn [prev] (drop_n is_multiple 6 prev))
	\7 (fn [prev] (drop_n is_multiple 7 prev))
	\8 (fn [prev] (drop_n is_multiple 8 prev))
	\9 (fn [prev] (drop_n is_multiple 9 prev))
})

(with-local-vars [line (read-line)]
	(while (not (= @line nil))
		(let [g (reduce (fn [s e] ((get f e) s)) (generate) (seq @line))]
		 	(with-local-vars [i 0]
				(while (< @i 10)
					(if (> @i 0) (print ","))
					(print (g))
					(var-set i (+ @i 1))
				)
				(println)
			)
		)
		(var-set line (read-line))
	)
)
